src/puppetlabs/puppetdb/middleware.clj

```diff
@@ -299,6 +299,34 @@
           (app (assoc req :body-string payload))
           (http/error-response (str "Missing required parameter 'payload'")))))))
 
+(defn max-command-size
+  "Returns the max command size relative to the current max heap. This
+  number was reached through testing of large catalogs and 1/205 was
+  the largest catalog that could be processed without GC or out of
+  memory errors"
+  []
+  (-> (Runtime/getRuntime)
+      .maxMemory
+      (/ 205)))
+
+(defn fail-when-payload-too-large
+  "Middlware that will return a 413 failure when the content-length of
+  a POST is too large (more than half the max heap). Acts as a noop
+  when that is not the case"
+  [app]
+  (let [max-content-length (max-command-size)]
+    (fn [req]
+      (if (and
+           (= :post (:request-method req))
+           (> (request/content-length req)
+              max-content-length))
+        (do
+          (.close (:body req))
+          {:status 413
+           :headers {}
+           :body "POST body too large"})
```

Comment:
That is to say: the certname is the main actionable piece of information here, since that'll help the user narrow down the source of the large catalog. Putting it in a a header would obliviate the need to deserialize the message; we could also do a single streaming parse over the content to just pull out the certname field. Jackson has a SAX-like event-based mode. Which is *tons* of fun to use.